### PR TITLE
Introduce Scope into semantic analyzer

### PR DESCRIPTION
**Pull request by kokosing**
Introduce Scope into semantic analyzer
    
    Scope is an object which contains information about entities available
    for current SQL clauses. It contains current RelationType as well as
    information about outer queries RelationTypes and available named queries.
    
    Semantic analyzer was changed in a way that each visit method gets and
    returns Scope objects. The input scope gives information about the context of
    given node. Contains information about outer relation types and
    accessible RelationType.
    Returned Scope contains information about produced entities
    (RelationType).
    
    Scopes are stored in Analysis per AST nodes.
    Each to find a Scope for a AST node (expression) we find the closest
    parent which has Scope set in Analysis.

### CODE REVIEW

**Comment by kokosing** [comment]
Superseded https://github.com/prestodb/presto/pull/5302

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java [hunk 1]
```diff
@@ -302,49 +293,34 @@ protected Type visitSymbolReference(SymbolReference node, StackableAstVisitorCon
...
-                    return field.getType();
-                }
-
-                if (!isColumnPrefix(qualifiedName, tupleDescriptor)) {
-                    if (isReferenceToOuterRelation(context.getContext(), qualifiedName)) {
-                        throw new SemanticException(NOT_SUPPORTED, node, "Correlated queries not yet supported. Invalid column reference: '%s'", qualifiedName);
-                    }
-                    throw createMissingAttributeException(node);
+                if (!scope.isColumnReference(qualifiedName)) {
+                    throw createMissingAttributeException(node, qualifiedName);
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: Why not `throwMissingAttributeException`? Exceptions have their stack traces filled during creation, not during throw. Creating an exception without throwing it immediately can lead to misleading stack traces in logs.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+    public Optional<Scope> getLateralScope()
+    {
+        return lateralScope;
+    }
+
+    public static final class Builder
+    {
+        private RelationType relationType = new RelationType();
+        private Optional<Boolean> approximate = Optional.empty();
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: Optional<Boolean> looks suspicious. In this case we don't use the third (empty) value of this tri-state logic, we just fall back to false upon build. Why not handle this using a field initializer?

`boolean approximate = false`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java (hunk 1 above)
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: Now I've seen the `SemanticExceptions` class and usage with `orElseThrow`. I'd consider migrating to `throwMissingAttributeException` and doing `optional.orElseThrow(() -> throwMissingAttributeException(...))` or  `optional.orElseGet(() -> throwMissingAttributeException(...))`, but I've got no strong preference to any of the ways (it's fine for me if it stays with `create...`)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 3]
```diff
@@ -442,33 +456,32 @@ else if (expressionType instanceof MapType) {
...
     {
         if (!table.getName().getPrefix().isPresent()) {
             // is this a reference to a WITH query?
             String name = table.getName().getSuffix();
 
-            WithQuery withQuery = context.getNamedQuery(name);
-            if (withQuery != null) {
-                Query query = withQuery.getQuery();
+            Optional<WithQuery> withQuery = scope.getNamedQuery(name);
+            if (withQuery.isPresent()) {
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Replace with `Optional.ifPresent(Consumer<T>)`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -442,33 +456,32 @@ else if (expressionType instanceof MapType) {
...
+                Query query = withQuery.get().getQuery();
                 analysis.registerNamedQuery(table, query);
 
                 // re-alias the fields with the name assigned to the query in the WITH declaration
                 RelationType queryDescriptor = analysis.getOutputDescriptor(query);
 
                 List<Field> fields;
-                if (withQuery.getColumnNames().isPresent()) {
+                Optional<List<String>> columnNames = withQuery.get().getColumnNames();
+                if (columnNames.isPresent()) {
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 5]
```diff
@@ -558,32 +567,28 @@ protected RelationType visitTable(Table table, AnalysisContext context)
...
-    protected RelationType visitAliasedRelation(AliasedRelation relation, AnalysisContext context)
+    protected Scope visitAliasedRelation(AliasedRelation relation, Scope scope)
     {
-        RelationType child = process(relation.getRelation(), context);
+        Scope relationScope = process(relation.getRelation(), scope);
 
         // todo this check should be inside of TupleDescriptor.withAlias, but the exception needs the node object
         if (relation.getColumnNames() != null) {
-            int totalColumns = child.getVisibleFieldCount();
+            int totalColumns = relationScope.getRelationType().getVisibleFieldCount();
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: extract variable for `relationScope.getRelationType()` (used below as well)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 6]
```diff
@@ -617,76 +622,70 @@ protected RelationType visitSampledRelation(SampledRelation relation, AnalysisCo
...
             if (outputFieldSize != descFieldSize) {
                 throw new SemanticException(MISMATCHED_SET_COLUMN_TYPES,
                         node,
                         "%s query has different number of fields: %d, %d",
                         setOperationName, outputFieldSize, descFieldSize);
             }
-            for (int i = 0; i < descriptor.getVisibleFields().size(); i++) {
-                Type descFieldType = descriptor.getFieldByIndex(i).getType();
+            for (int i = 0; i < descFieldSize; i++) {
+                Type descFieldType = relationScope.getRelationType().getFieldByIndex(i).getType();
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: extract variable `relationScope.getRelationType()` (used below in loop)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 7]
```diff
@@ -699,62 +698,57 @@ protected RelationType visitSetOperation(SetOperation node, AnalysisContext cont
...
         }
-        RelationType outputDescriptor = new RelationType(outputDescriptorFields);
-        analysis.setOutputDescriptor(node, outputDescriptor);
 
         for (int i = 0; i < node.getRelations().size(); i++) {
             Relation relation = node.getRelations().get(i);
-            RelationType descriptor = descriptors[i];
-            for (int j = 0; j < descriptor.getVisibleFields().size(); j++) {
+            Scope relationScope = relationScopes.get(i);
+            for (int j = 0; j < relationScope.getRelationType().getVisibleFields().size(); j++) {
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: extract relationType

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 8]
```diff
@@ -699,62 +698,57 @@ protected RelationType visitSetOperation(SetOperation node, AnalysisContext cont
...
 
     @Override
-    protected RelationType visitExcept(Except node, AnalysisContext context)
+    protected Scope visitExcept(Except node, Scope scope)
     {
         throw new SemanticException(NOT_SUPPORTED, node, "EXCEPT not yet implemented");
     }
 
     @Override
-    protected RelationType visitJoin(Join node, AnalysisContext context)
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Would be great to split this method into smaller ones, but I think we can do that after this gets merged to avoid confusion during review

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 9]
```diff
@@ -1286,7 +1277,7 @@ else if (expression instanceof DereferenceExpression) {
                 // expand * and T.*
                 Optional<QualifiedName> starPrefix = ((AllColumns) item).getPrefix();
 
-                List<Field> fields = tupleDescriptor.resolveFieldsWithPrefix(starPrefix);
+                List<Field> fields = scope.getRelationType().resolveFieldsWithPrefix(starPrefix);
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: extract relationType

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestScope.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertTrue;
+
+public class TestScope
+{
+    @Test
+    public void test()
```
Comment (ArturGajowy) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestScope.java]: Extract test cases with names describing expected scope properties

**Comment by ArturGajowy** [comment]
Looks yummy, except the TestScope#test that needs some love, and the other minor comments.

**Comment by ArturGajowy** [comment]
Let me add that I really like how the `Scope` concept emerges from AnalysisContext and is not an anemic DTO.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalyzer.java]: thanks, done

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java (hunk 2 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: In case of empty `approximate` its value is inherited from scope parent. However I used your suggestion for `queryBoundary`

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 3 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: I cannot as I cannot return from the `visitTable` from that lambda

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: ditto ;)

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 8 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: That method changed a lot recently. I would not touch it in this PR if it is not needed.

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestScope.java (hunk 10 above)
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestScope.java]: well... this is only a single test scenario
Extracting multiple test cases would create a lot of boilerplate code

**Comment by kokosing** [comment]
Rebased and applied comments.

**Comment by martint** [comment]
Merged 1e320b8 and c3a075f

**Comment by kokosing** [comment]
@martint Thanks! Sorry about these two commits contained bugs. I rebased the PR.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
+            return this;
+        }
+
+        public Builder withParent(Scope parent)
+        {
+            this.parent = Optional.of(parent);
+            return this;
+        }
+
+        public Builder withQueryBoundary(boolean queryBoundary)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Scope.java]: Is this ever called with `false`? I don't think there's any use case for that.

I'd change this method to take no arguments and rename it to `markQueryBoundary` or `makeQueryBoundary`, or maybe even have a different `builder()` method in `Scope` to create a nested scope.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java [hunk 12]
```diff
@@ -59,8 +59,7 @@ public Analysis analyze(Statement statement)
         Statement rewrittenStatement = StatementRewrite.rewrite(session, metadata, sqlParser, queryExplainer, statement);
         Analysis analysis = new Analysis(rewrittenStatement);
         StatementAnalyzer analyzer = new StatementAnalyzer(analysis, metadata, sqlParser, accessControl, session, experimentalSyntaxEnabled);
-        RelationType outputDescriptor = analyzer.process(rewrittenStatement, new AnalysisContext());
-        analysis.setOutputDescriptor(outputDescriptor);
+        analyzer.process(rewrittenStatement, Scope.builder().build());
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java]: For consistency, should this scope be a query boundary?

File: presto-parser/src/main/java/com/facebook/presto/sql/tree/DefaultTraversalVisitor.java [hunk 13]
```diff
@@ -498,4 +498,13 @@ protected R visitCreateTableAsSelect(CreateTableAsSelect node, C context)
 
         return null;
     }
+
+    @Override
+    protected R visitExplain(Explain node, C context)
```
Comment (martint) [inline presto-parser/src/main/java/com/facebook/presto/sql/tree/DefaultTraversalVisitor.java]: Unrelated change? Move to a separate commit.

**Comment by martint** [comment]
Just a couple of minor comments, but otherwise looks pretty good.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java (hunk 12 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analyzer.java]: Good point.

**Comment by kokosing** [comment]
@martint Rebased and addressed comments

**Comment by martint** [comment]
Merged, thanks!
